src/window/sound_options.c

```diff
@@ -144,12 +144,16 @@ static void button_ok(int param1, int param2)
 
 static void button_cancel(int param1, int param2)
 {
+    int music_sound_is_dirty = setting_sound_is_enabled(SOUND_MUSIC) != data.original_music.enabled;
+
     setting_reset_sound(SOUND_EFFECTS, data.original_effects.enabled, data.original_effects.volume);
     setting_reset_sound(SOUND_MUSIC, data.original_music.enabled, data.original_music.volume);
     setting_reset_sound(SOUND_SPEECH, data.original_speech.enabled, data.original_speech.volume);
     setting_reset_sound(SOUND_CITY, data.original_city.enabled, data.original_city.volume);
     if (data.original_music.enabled) {
-        sound_music_reset();
+        if(music_sound_is_dirty){
+            sound_music_reset();
+        }
         sound_music_update();
```

Comment:
`sound_music_update();` can also be placed within the `if` above: this function recalculates which track to play -- that cannot have changed since opening this dialog.